source/str_util.c

```diff
@@ -30,47 +30,828 @@
 #define ENC_WIDE 2
 #define ENC_NARROW 3
 
-static int widths_len = 2*38;
+static int widths_len = 2*820;
```

Comment:
Doing a linear search over 820 items might slow. Should we think of using a more efficient data structure or am I overthinking this?

Reply:
I'm even thinking about depreciation of C code and migration on `wcwidth.
It became global standard, handle multiple unicode versions and since author allow to copy-paste unicode version detect logic - we can better handle different terminals.
Re-implementation of unicode tables in C or migration on different language (like rust) at this moment is too painful

Reply:
detecting things about the terminal is tricky because the text calculations are separated from the screen instance. If someone is connecting to an urwid app over ssh like the twisted ssh example how do we know which unicode version detection to use? (there can be many users attached each with their own screen)

Reply:
but, using an existing library for the width calculation sounds like a great idea

Reply:
And simple copy-paste of logic and content (even if separate generated) looks strange.
We can use binary search, it is faster.

Reply:
I checked: for current release we can use binary search and later we can migrate C part to rust and use upstream libraries for char width both in python and binary.

Reply:
Re: use of wcwidth - good idea to do so, but I suggest you have a look at a recent issue posted by the author on his github, where he discusses the limits of UNICODE_VERSION https://github.com/jquast/wcwidth/issues/104

the tl dr is that it's challenging to fix unicode version compatibility issues at the app layer, and the way forward is to get terminal emulator authors to stay up to date with the latest unicode version.

Reply:
> Re: use of wcwidth - good idea to do so, but I suggest you have a look at a recent issue posted by the author on his github, where he discusses the limits of UNICODE_VERSION https://github.com/jquast/wcwidth/issues/104
>

I fully understand that issue, but it also has limitations: in case of widget decorations, incorrect symbols length of decorated widget will produce incorrect decorations length.
Technically we can play "stupid" and go forward with the last unicode version tables, since only few terminals at this moment fully stuck at unicode 12.

> the tl dr is that it's challenging to fix unicode version compatibility issues at the app layer, and the way forward is to get terminal emulator authors to stay up to date with the latest unicode version.

Reply:
Understood. Makes sense to update the tables. I am curious, why update to v.12 and not the latest v.15.1?

Reply:
Originally went to the most supported.
After rethinking on weekends went to the 15.1
Technically we have only 1 fully working solution: ask terminal for cursor position after each double width character and add space if it rendered single width, but this solution is extremely expensive and works only with RawDisplay.